File: 122-best-time-to-buy-and-sell-stock-ii/best-time-to-buy-and-sell-stock-ii.cpp
class Solution {
private:
    long long NEG_INF = -1e14;

public:
    long long maxProfit(vector<int>& prices) {
        long long cash = 0;
        long long hold = NEG_INF;

        for (int p : prices) {
            long long new_cash = std::max(cash, hold + p);
            long long new_hold = std::max(hold, cash - p);

            cash = new_cash;
            hold = new_hold;
        }
        return cash;
    }
};

File: 122-best-time-to-buy-and-sell-stock-ii/best-time-to-buy-and-sell-stock-ii.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # buy at valley, sell at peak
        profit = 0
        i = 0
        n = len(prices) - 1

        while i < n:
            while i < n and prices[i] >= prices[i+1]:
                i += 1
            buy = prices[i]
            
            while i < n and prices[i] <= prices[i+1]:
                i += 1
            sell = prices[i]

            profit += sell - buy
        
        return profit